modules/packages/Channel.chpl

```diff
@@ -0,0 +1,605 @@
+                recvWaiters.enqueue(processing);
+                unlock();
+                var status = processing.suspend();
+                delete processing;
+                return status;
+            }
+
+            if bufferSize > 0 {
+                val = buffer[recvIdx];
+            }
+
+            if !sendWaiters.isEmpty() {
+                var sender = sendWaiters.dequeue();
+                if bufferSize > 0 {
+                    buffer[recvIdx] = sender.val.deref();
+
+                    sendIdx = (sendIdx + 1) % bufferSize;
+                    recvIdx = (recvIdx + 1) % bufferSize;
+                }
+                else val = sender.val.deref();
+
+                sender.notify(true);
+            }
+            else {
+
+                recvIdx = (recvIdx + 1) % bufferSize;
+                count -= 1;
+
+            }
+            if blocking then unlock();
+
+            return true;
+
+        }
+
+        proc send(in val : eltType, blocking : bool) : bool throws {
+            if blocking then lock();
+
+            if closed {
+                if blocking then unlock();
+                throw new owned ChannelError("Trying to send on a closed channel");
+            }
+
+            while !recvWaiters.isEmpty() && recvWaiters.front!.isSelect {
+                if recvWaiters.front!.isSelectDone.deref().compareAndSwap(-1, recvWaiters.front!.selectId) {
+                    break;
+                }
+                else recvWaiters.dequeue();
+            }
+
+            if count == bufferSize && recvWaiters.isEmpty() {
+                if !blocking then return false;
+                var process$ : single bool;
+                var processing = new unmanaged Waiter(val, process$);
+
+                sendWaiters.enqueue(processing);
+                unlock();
+                var status = processing.suspend();
+                delete processing;
+                if status == false {
+                    throw new owned ChannelError("Trying to send on a closed channel");
+                }
+                return status;
+
+            }
+            else {
+                if !recvWaiters.isEmpty() {
+                    var receiver = recvWaiters.dequeue();
+                    receiver.val.deref() = val;
+
+                    receiver.notify(true);
+                }
+                else {
+                    buffer[sendIdx] = val;
+
+                    sendIdx = (sendIdx + 1) % bufferSize;
+                    count += 1;
+                }
+
+                if blocking then unlock();
+                return true;
+            }
+        }
+
+        proc close() throws {
+
+            lock();
+            if closed {
+                unlock();
+                throw new owned ChannelError("Trying to close a closed channel");
+            }
+            closed = true;
+            var queued = new WaiterQueue(eltType);
+            while(!recvWaiters.isEmpty()) {
```

Comment:
you haven't used `()` for while/ifs so far. make sure that you're consistent.

My preference would be no `()` throughout, but I'm OK as long as you're consistent.